File: src/main/java/no/novari/flyt/file/FileRepository.java
package no.novari.flyt.file;

import lombok.extern.slf4j.Slf4j;
import no.novari.flyt.AzureBlobAdapter;
import org.springframework.stereotype.Repository;
import reactor.core.publisher.Mono;

import java.util.List;
import java.util.Optional;
import java.util.UUID;

@Repository
@Slf4j
public class FileRepository {

    private final AzureBlobAdapter azureBlobAdapter;

    public FileRepository(AzureBlobAdapter azureBlobAdapter) {
        this.azureBlobAdapter = azureBlobAdapter;
    }

    public Mono<UUID> putFile(UUID fileId, File file) {
        return azureBlobAdapter.uploadFile(fileId, file)
                .doOnNext(response -> logSuccessfulAction(fileId, "uploaded"));
    }

    public Mono<Optional<File>> findById(UUID fileId) {
        return azureBlobAdapter.downloadFile(fileId)
                .doOnNext(optionalFile -> {
                    if (optionalFile.isPresent()) {
                        logSuccessfulAction(fileId, "found");
                    } else {
                        logUnsuccessfulAction(fileId, "find");
                    }
                })
                .doOnError(e -> log.error("Could not download file", e));
    }

    public Mono<Void> deleteFiles(List<UUID> fileIds) {
        return azureBlobAdapter.deleteFilesByIds(fileIds)
                .doOnSuccess(aVoid -> fileIds.forEach(fileId -> logSuccessfulAction(fileId, "deleted")))
                .doOnError(e -> log.error("Could not delete files", e));
    }

    public int deleteFilesOlderThan(int days) {
        List<DeletedFile> deletedFiles = azureBlobAdapter.deleteFilesOlderThanDays(days);
        deletedFiles
                .forEach(deletedFile ->
                        log.info("deleted file with name {}, timestamp {}", deletedFile.name(), deletedFile.deletedAt())
                );
        return deletedFiles.size();
    }

    private void logSuccessfulAction(UUID fileId, String performedAction) {
        log.info("Successfully {} File{fileId={}} in Azure Blob Storage", performedAction, fileId);
    }

    private void logUnsuccessfulAction(UUID fileId, String actionToBePerformed) {
        log.warn("Could not {} File{fileId={}} in Azure Blob Storage", actionToBePerformed, fileId);
    }

}

File: src/main/kotlin/no/novari/flyt/files/infrastructure/storage/FileRepository.kt
package no.novari.flyt.files.infrastructure.storage

import io.github.oshai.kotlinlogging.KotlinLogging
import no.novari.flyt.files.domain.FilePayload
import no.novari.flyt.files.domain.exception.FileStorageException
import org.springframework.stereotype.Repository
import java.util.UUID

@Repository
class FileRepository(
    private val blobStorageAdapter: BlobStorageAdapter,
) {
    private val log = KotlinLogging.logger {}

    fun putFile(
        fileId: UUID,
        file: FilePayload,
    ): UUID {
        return try {
            val uploadedFileId = blobStorageAdapter.uploadFile(fileId, file)
            logSuccessfulAction(fileId, "uploaded")
            uploadedFileId
        } catch (exception: Exception) {
            log.atError {
                message = "Could not upload file with fileId={}"
                arguments = arrayOf(fileId)
                cause = exception
            }
            throw FileStorageException("Could not upload file", exception)
        }
    }

    fun findById(fileId: UUID): FilePayload? {
        return try {
            val filePayload = blobStorageAdapter.downloadFile(fileId)
            if (filePayload != null) {
                logSuccessfulAction(fileId, "found")
            } else {
                logUnsuccessfulAction(fileId, "find")
            }
            filePayload
        } catch (exception: Exception) {
            log.atError {
                message = "Could not download file with fileId={}"
                arguments = arrayOf(fileId)
                cause = exception
            }
            throw FileStorageException("Could not download file", exception)
        }
    }

    fun deleteFiles(fileIds: List<UUID>) {
        try {
            blobStorageAdapter.deleteFilesByIds(fileIds)
            fileIds.forEach { fileId -> logSuccessfulAction(fileId, "deleted") }
        } catch (exception: Exception) {
            log.atError {
                message = "Could not delete files with fileIds={}"
                arguments = arrayOf(fileIds)
                cause = exception
            }
            throw FileStorageException("Could not delete files", exception)
        }
    }

    fun deleteFilesOlderThan(days: Int): Int {
        try {
            val deletedFiles = blobStorageAdapter.deleteFilesOlderThanDays(days)
            deletedFiles.forEach { deletedFile ->
                log.atInfo {
                    message = "deleted file with name {}, timestamp {}"
                    arguments = arrayOf(deletedFile.name, deletedFile.deletedAt)
                }
            }
            return deletedFiles.size
        } catch (exception: Exception) {
            log.atError {
                message = "Could not delete files older than {} days"
                arguments = arrayOf(days)
                cause = exception
            }
            throw FileStorageException("Could not delete old files", exception)
        }
    }

    private fun logSuccessfulAction(
        fileId: UUID,
        performedAction: String,
    ) {
        log.atInfo {
            message = "Successfully {} File{fileId={}} in file storage"
            arguments = arrayOf(performedAction, fileId)
        }
    }

    private fun logUnsuccessfulAction(
        fileId: UUID,
        actionToBePerformed: String,
    ) {
        log.atWarn {
            message = "Could not {} File{fileId={}} in file storage"
            arguments = arrayOf(actionToBePerformed, fileId)
        }
    }
}
